crates/bevy_ecs/src/query/access.rs

```diff
@@ -325,6 +318,84 @@ impl<T: SparseSetIndex> FilteredAccess<T> {
     }
 }
 
+// A struct to express something like `Or<(With<A>, With<B>)>`.
+// Filters like `(With<A>, Or<(With<B>, With<C>)>` are expanded into `Or<((With<A>, With<B>), (With<A>, With<C>))>`.
+#[derive(Clone, Eq, PartialEq)]
+struct ExpandedOrAccess<T, I> {
+    arr: SmallVec<[FixedBitSet; 8]>,
+    // To prevent mixing `With` and `Without` filters.
+    _filter_type: PhantomData<T>,
+    _index_type: PhantomData<I>,
+}
+
+#[derive(Debug, Clone, Eq, PartialEq)]
+struct WithMarker;
+
+#[derive(Debug, Clone, Eq, PartialEq)]
+struct WithoutMarker;
+
+impl<T, I: SparseSetIndex + fmt::Debug> fmt::Debug for ExpandedOrAccess<T, I> {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        f.debug_list()
+            .entries(self.arr.iter().map(FormattedBitSet::<I>::new))
+            .finish()
+    }
+}
+
+impl<T, I: SparseSetIndex> Default for ExpandedOrAccess<T, I> {
+    fn default() -> Self {
+        Self {
+            arr: smallvec::smallvec![FixedBitSet::default()],
+            _filter_type: PhantomData,
+            _index_type: PhantomData,
+        }
+    }
+}
+
+impl<T, I: SparseSetIndex> ExpandedOrAccess<T, I> {
+    fn add(&mut self, index: usize) {
+        for with in &mut self.arr {
+            with.grow(index + 1);
+            with.insert(index);
+        }
+    }
+
+    fn extend_with_or(&mut self, other: &Self) {
+        self.arr.append(&mut other.arr.clone());
+    }
+
+    fn union_with(&mut self, other: &Self) {
+        if other.arr.len() == 1 {
+            for with in &mut self.arr {
+                with.union_with(&other.arr[0]);
+            }
+            return;
+        }
+
+        let mut new_with = SmallVec::with_capacity(self.arr.len() * other.arr.len());
+        for with in &self.arr {
+            for other_with in &other.arr {
+                let mut w = with.clone();
+                w.union_with(other_with);
+                new_with.push(w);
+            }
+        }
+        self.arr = new_with;
+    }
+}
+
+impl<I: SparseSetIndex> ExpandedOrAccess<WithMarker, I> {
+    fn is_disjoint(&self, other: &ExpandedOrAccess<WithoutMarker, I>) -> bool {
+        // We need to make sure that every `without` variant is disjoint with any `with`.
+        // For example, `Or<(Without<A>, Without<B>)>` is disjoint with `Or<(With<A>, With<B>)>`,
```

Comment:
This doesn't looks correct to me. Both `Or<(Without<A>, Without<B>)>` and `Or<(With<A>, With<B>)>` match entities that have exactly one component among `A` and `B` but not both, so they're not completly disjoint.

Reply:
Yeah, I think you're right, that doesn't fully make sense. "Disjoint" is probably a poor word choice, but what I meant is that for the compatibility check it matters only if every `without` AND sequence is disjoint with at least one `with` AND sequence (in which case we can't guarantee that the queries are compatible).

(I'm trying to argue that the comment is correct in the context of how the method is used, apart from the fact that "disjoint" is a bad term here.)

Will it help if we rename the method to something like `is_ruled_out_by` and negate the returned result? Then the semantics would be
```rust
pub fn is_compatible(&self, other: &FilteredAccess<T>) -> bool {
    self.access.is_compatible(&other.access)
        || self.with.is_ruled_out_by(&other.without)
        || other.with.is_ruled_out_by(&self.without)
}
```

I hope it'll help to paraphrase the comment to be less confusing.

Reply:
> Will it help if we rename the method to something like `is_ruled_out_by` and negate the returned result?

`is_ruled_out_by` looks like a better term to me. Disjoint is definitely confusing because the `with`/`without` AND sequences being disjoint mean the opposite of the queries being disjoint.

> it matters only if every `without` AND sequence is disjoint with at least one `with` AND sequence

This seems flawed to me. What if not every `without` AND sequence is disjoint with at least one `with` AND sequence, but at least one is? Then you still get one set of possible, which means the two queries are not actually compatible.

Actually though I think looking at them separately is wrong. See my comment below.

Reply:
Fixed and added the tests covering the previously failing scenarios